models/vrp.rb

```diff
@@ -452,6 +452,30 @@ def self.deduce_unavailable_days(hash, element, start_index, end_index, type)
       collect_unavaible_day_indices(element, start_index, end_index)
     end
 
+    def self.deduce_possible_days(hash, element, start_index)
+      convert_possible_dates_into_indices(element, hash, start_index)
+
+      element[:first_possible_days] = element[:first_possible_day_indices].to_a.slice(0..(element[:visits_number] || 1) - 1)
+      element[:last_possible_days] = element[:last_possible_day_indices].to_a.slice(0..element[:visits_number] || 1 - 1)
+
+      %i[first_possible_day_indices first_possible_dates last_possible_day_indices last_possible_dates].each{ |k|
+        element.delete(k)
+      }
+    end
+
+    def self.convert_possible_dates_into_indices(element, hash, start_index)
+      return unless hash[:configuration][:schedule][:range_date]
+
+      start_value = hash[:configuration][:schedule][:range_date][:start].to_date.ajd.to_i - start_index
+      element[:first_possible_day_indices] ||= element[:first_possible_dates].to_a.collect{ |date|
+        date.to_date.ajd.to_i - start_value
```

Comment:
This corresponds to astronomical Julian day. I used same logic as we originally had. We should discuss it with @braktar, you can create a ticket if you think we should do this differently.

Reply:
> This corresponds to astronomical Julian day. 

I already check the description of that function but I couldn't understand the **significance** of Astronomical Julian day that's why I asked if you could elaborate. What does it do and why not any other functions instead ( `ld` or `jd` for example, which already return integers in the first place, instead of a rational number) or why not we just directly do `(date.to_date - start_date).to_i` (`Date` class implements simple arithmetic operations like `-` and `+`) and Ruby handles the gritty details.

> I used same logic as we originally had.

Oh so we already have the same logic for the other "date -> index" conversion, is that it? If that is the case, can you extract the logic to a function so that we are sure that all such conversion are done via the same `convert_date_to_index(date, range_start)` function. So that there will be only one conversion logic.

> We should discuss it with @braktar, you can create a ticket if you think we should do this differently.

Sure, you can extract the `date.to_date.ajd.to_i - start_value` to a `convert_date_to_index` function and I will create a ticket to investigate if we need `ajd` or not. I am just afraid of messing up date logic because such little details can have unintended consequences -- for example [Twitter had crashed](https://www.theguardian.com/technology/2014/dec/29/twitter-2015-date-bug) for android users because a date formatted in ISO year format (%G) instead of usual/ordinary year format (%Y). Since I didn't understand the significance of Astronomical Julian day and since I don't know why we just didn't do `(date - date).to_i` in the first place, I thought I should ask ;)